### ISSUE TITLE
Jackson ignores float deserialization feature when custom field converter is used

### ISSUE THREAD

**Issue by Gwulior** [FasterXML/jackson-databind#2047]
Hi there, was faced with a problem, looks like I follow manual but it does not work.
I have class with field that can have different structure, so I use 
```java
    @JsonDeserialize(using = ClientDataFieldDeserializer.class)
    private Object data;
```
I want to use BigDecimal so I followed this instruction 
https://github.com/FasterXML/jackson-databind/wiki/Deserialization-Features

But when I perform

```java
ObjectReader reader = objectMapper.reader(clazz); (HashMap class there)
            if (feature != null) {
                reader = reader.with(feature);
            }
            return reader.readValue(data);
```

It completely ignores the enabled feature.

The intresting thing is when I debug, I can see in converter 

```java
  @Override
    public Object deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
        JsonToken jsonToken = p.getCurrentToken();
        Object result;
        if (jsonToken == JsonToken.VALUE_STRING) {
            result = p.getValueAsString();
        } else if (jsonToken == JsonToken.START_OBJECT) {
            result = p.readValueAs(HashMap.class);
        } else {
            result = null;
        }
        return result;
    }
```
That DeserializationContext ctxt has this feature enabled, but when we stop in Jackson sources UntypedObjectDeserializer, 472 line
I can see that this code
```java
case JsonTokenId.ID_NUMBER_FLOAT:
                if (ctxt.isEnabled(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS)) {
                    return jp.getDecimalValue();
                }
                return Double.valueOf(jp.getDoubleValue());
```
can not see enabled feature. So finally I managed to solve it only enabled the feature globally on ObjectMapper.

**Comment by cowtowncoder** [comment]
I am sorry but I do not quite understand what is your exact problem here.
Could you write a unit test showing what you expect to work, but does not, failing an assertion?

**Comment by Gwulior** [comment]
Here you are
https://github.com/Gwulior/jackson-float-bug

**Comment by cowtowncoder** [comment]
Thank you!

**Comment by cowtowncoder** [comment]
Ah. This is an interesting problem, and something that with code you have will only be fixed for 3.0.
The issue is that although `JsonParser.readValueAs(...)` is convenient, it only has access to `ObjectMapper` that parser was created with. But not to possibly different configuration that `ObjectReader` has. As a result it does not see changes reader has.
There is no way to fix that with 2.x, but will work as expected in 3.0.

However. You can improve your deserializer by using `readValue()` methods from `DeserializationContext`. That is preferred way anyway since it incurs less overhead, but it also retains context with same configuration.
This resolves your issue.
